### PR TITLE
Added support to read diagnostics config from the path defined in environment variable

### PR DESCRIPTION
**Pull request by Yun-Ting**
Fix Issue https://github.com/microsoft/ApplicationInsights-dotnet/issues/2724.

## Changes

Design assumption:
Function Apps environment will set the value of this environment variable: APPLICATIONINSIGHTS_LOG_DIAGNOSTICS to be a directory that
1. Users can create/delete diagnostics file and,
2. SDK has the privilege to write to
at host initialization.

The SDK reads the value of APPLICATIONINSIGHTS_LOG_DIAGNOSTICS, and parse settings of ``ApplicationInsightsDiagnostics.json`` file and write diagnostics log to the directory.

### Checklist
- [x] I ran Unit Tests locally.
- [ ] CHANGELOG.md updated with one line description of the fix, and a link to the original issue if available.

The PR will trigger build, unit tests, and functional tests automatically. Please follow [these](https://github.com/Microsoft/ApplicationInsights-dotnet/blob/develop/.github/CONTRIBUTING.md) instructions to build and test locally.

### Notes for authors:
- FxCop and other analyzers will fail the build. To see these errors yourself, compile localy using the Release configuration.

### CODE REVIEW

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigParser.cs [hunk 1]
```diff
@@ -40,8 +43,14 @@ public bool TryGetConfiguration(out string logDirectory, out int fileSizeInKB, o
             {
                 var configFilePath = ConfigFileName;
 
+                if (PlatformSingleton.Current.TryGetEnvironmentVariable(LogDiagnosticsEnviornmentVariable, out string logDiagnosticsPath))
+                {
+                    configFilePath = Path.Combine(logDiagnosticsPath, ConfigFileName);
+                    logDirectory = logDiagnosticsPath;
+                }
+
```
Comment (xiang17) [inline BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigParser.cs:51]: remove the blank space before `else if`.

File: BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigRefresherTest.cs [hunk 2]
```diff
@@ -74,6 +77,52 @@ public void SelfDiagnosticsConfigRefresher_CaptureAsConfigured()
...
 
+        [TestMethod]
+        public void SelfDiagnosticsConfigRefresher_ReadFromEnviornmentVar()
+        {
+
+            var key = "APPLICATIONINSIGHTS_LOG_DIAGNOSTICS";
+            var val = string.Empty;
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                val = "C:\\home\\LogFiles\\SelfDiagnostics";
```
Comment (reyang) [inline BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigRefresherTest.cs:88]: maybe use https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/verbatim

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigParser.cs [hunk 3]
```diff
@@ -5,13 +5,16 @@
...
     using System.Text.RegularExpressions;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Platform;
 
     internal class SelfDiagnosticsConfigParser
     {
         public const string ConfigFileName = "ApplicationInsightsDiagnostics.json";
         private const int FileSizeLowerLimit = 1024;  // Lower limit for log file size in KB: 1MB
         private const int FileSizeUpperLimit = 128 * 1024;  // Upper limit for log file size in KB: 128MB
 
+        private const string LogDiagnosticsEnviornmentVariable = "APPLICATIONINSIGHTS_LOG_DIAGNOSTICS";
```
Comment (reyang) [inline BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigParser.cs:16]:
```suggestion
        private const string LogDiagnosticsEnvironmentVariable = "APPLICATIONINSIGHTS_LOG_DIAGNOSTICS";
```

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigParser.cs [hunk 4]
```diff
@@ -40,8 +43,14 @@ public bool TryGetConfiguration(out string logDirectory, out int fileSizeInKB, o
             {
                 var configFilePath = ConfigFileName;
 
+                if (PlatformSingleton.Current.TryGetEnvironmentVariable(LogDiagnosticsEnviornmentVariable, out string logDiagnosticsPath))
+                {
+                    configFilePath = Path.Combine(logDiagnosticsPath, ConfigFileName);
+                    logDirectory = logDiagnosticsPath;
+                }
+
                 // First check using current working directory
```
Comment (xiang17) [inline BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigParser.cs:52]:
```suggestion
                // Second check using current working directory
```

Update the comments for all three checks accordingly.

File: BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigRefresherTest.cs [hunk 5]
```diff
@@ -91,13 +140,25 @@ private static byte[] ReadFile(string filePath, int byteCount)
             }
         }
 
-        private static void CreateConfigFile()
+        private static void CreateConfigFile(bool userDefinedLogDirectory = true)
```
Comment (xiang17) [inline BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigRefresherTest.cs:143]: Did you ever invoke this method with `false`?

**Review by xiang17 — APPROVED** [review]
LGTM

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigParser.cs (hunk 1 above)
Comment (Yun-Ting) [reply BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigParser.cs:52]: do we want to have a blank line before comment?

File: BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigRefresherTest.cs (hunk 5 above)
Comment (Yun-Ting) [reply BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigRefresherTest.cs:143]: updated testcase

File: BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigRefresherTest.cs (hunk 2 above)
Comment (Yun-Ting) [reply BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigRefresherTest.cs]: updated

File: BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigRefresherTest.cs [hunk 6]
```diff
@@ -74,6 +77,52 @@ public void SelfDiagnosticsConfigRefresher_CaptureAsConfigured()
...
+            var key = "APPLICATIONINSIGHTS_LOG_DIAGNOSTICS";
+            var val = string.Empty;
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                val = @"C:\home\LogFiles\SelfDiagnostics";
+            }
+            else
+            {
+                val = "/home/user/LogFiles/SelfDiagnostics";
+            }
```
Comment (reyang) [inline BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Extensibility/Implementation/Tracing/SelfDiagnostics/SelfDiagnosticsConfigRefresherTest.cs:93]: Minor: maybe simplify this to:

```suggestion
            var val = "/home/user/LogFiles/SelfDiagnostics";
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                val = @"C:\home\LogFiles\SelfDiagnostics";
            }
```

**Review by reyang — APPROVED** [review]

**Comment by cijothomas** [comment]
> Function Apps environment will set the value of this environment variable: APPLICATIONINSIGHTS_LOG_DIAGNOSTICS to be a directory that
> 
> Users can create/delete diagnostics file and,
> SDK has the privilege to write to
> at host initialization.
> The SDK reads the value of APPLICATIONINSIGHTS_LOG_DIAGNOSTICS, and parse settings of ApplicationInsightsDiagnostics.json file and write diagnostics log to the directory.

Bit confused with this description.
The following is my understanding:
The SDK needs read permission for the directory specified in APPLICATIONINSIGHTS_LOG_DIAGNOSTICS.
The SDK needs write permission for the directory specified in ApplicationInsightsDiagnostics.json, it could be different directory than specified in APPLICATIONINSIGHTS_LOG_DIAGNOSTICs.

**Review by cijothomas — APPROVED** [review]

**Review by TimothyMothra — APPROVED** [review]

**Comment by TimothyMothra** [comment]
@Yun-Ting please add an entry in the changelog.

**Pull request merged**